readthedocs/oauth/services/gitlab.py

```diff
@@ -0,0 +1,330 @@
+        :rtype: RemoteOrganization
+        """
+        try:
+            organization = RemoteOrganization.objects.get(
+                slug=fields.get('path'),
+                users=self.user,
+                account=self.account,
+            )
+        except RemoteOrganization.DoesNotExist:
+            organization = RemoteOrganization.objects.create(
+                slug=fields.get('path'),
+                account=self.account,
+            )
+            organization.users.add(self.user)
+
+        organization.name = fields.get('name')
+        organization.account = self.account
+        organization.url = '{url}/{path}'.format(
+            url=self.adapter.provider_base_url,
+            path=fields.get('path'),
+        )
+        avatar = fields.get('avatar') or {}
+        if avatar.get('url'):
+            organization.avatar_url = '{url}/{avatar}'.format(
+                url=self.adapter.provider_base_url,
+                avatar=avatar.get('url'),
+            )
+        else:
+            # TODO: do we want default avatar URL here?
+            organization.avatar_url = self.default_avatar['org']
+
+        organization.json = json.dumps(fields)
+        organization.save()
+        return organization
+
+    def get_webhook_data(self, repo_id, integration, project):
+        """
+        Get webhook JSON data to post to the API.
+
+        See: http://doc.gitlab.com/ce/api/projects.html#add-project-hook
+        """
+        return json.dumps({
+            'id': repo_id,
+            'push_events': True,
+            'tag_push_events': True,
+            'url': 'https://{domain}{path}'.format(
+                domain=settings.PRODUCTION_DOMAIN,
+                path=reverse(
+                    'api_webhook',
+                    kwargs={
+                        'project_slug': project.slug,
+                        'integration_pk': integration.pk,
+                    },
+                ),
+            ),
+
+            # Optional
+            'issues_events': False,
+            'merge_requests_events': False,
+            'note_events': False,
+            'job_events': False,
+            'pipeline_events': False,
+            'wiki_events': False,
+        })
+
+    def setup_webhook(self, project):
+        """
+        Set up GitLab project webhook for project.
+
+        :param project: project to set up webhook for
+        :type project: Project
+        :returns: boolean based on webhook set up success
+        :rtype: bool
+        """
+        session = self.get_session()
+        owner, repo = build_utils.get_gitlab_username_repo(url=project.repo)
+        integration, _ = Integration.objects.get_or_create(
+            project=project,
+            integration_type=Integration.GITLAB_WEBHOOK,
+        )
+
+        # The ID or URL-encoded path of the project
+        # https://docs.gitlab.com/ce/api/README.html#namespaced-path-encoding
+        repo_id = '{}%2F{}'.format(owner, repo)
```

Comment:
This is enough here but maybe it's better to find a way to get the proper `repo:id` from gitlab

Reply:
this is probably fine. Github gives us this information in the api return, but i don't see anything wrong with this as long as owner and repo are supplied from the API.

Reply:
I found a better way to do this,

```python
        # The ID or URL-encoded path of the project
        # https://docs.gitlab.com/ce/api/README.html#namespaced-path-encoding
        repo_id = json.loads(project.remote_repository.json).get('id')
```

It uses the RemoteRepository response saved in our database, which I think it's better than parsing a URL, more specifically in GitLab that the URL could have different domains.

So, I'll try to use the repo `id` here instead of owner + repo and we can get rid of the GL_REGEX also, since it won't be needed anymore.

Reply:
The regex is also used at this point:

https://github.com/rtfd/readthedocs.org/blob/4db7339aac7a52d7a35b6359eff8032e8aadf6ae/readthedocs/doc_builder/backends/sphinx.py#L72-L75

So, we will need to figure this out for that case, at least.